sbtc-core/src/signer/mod.rs

```diff
@@ -0,0 +1,119 @@
+/// An sBTC signer configuration
+pub mod config;
+
+use std::collections::HashMap;
+
+use crate::signer::config::Config;
+use crate::SBTCError;
+use bitcoin::{Network, PrivateKey, PublicKey, Transaction as BitcoinTransaction};
+use p256k1::ecdsa;
+use url::Url;
+use wsts::{bip340::SchnorrProof, common::Signature};
+
+#[derive(Default, Clone, Debug)]
+pub struct SignersData {
+    pub signers: HashMap<u32, ecdsa::PublicKey>,
+    pub vote_shares: HashMap<u32, ecdsa::PublicKey>,
+}
+
+/// Coordinator algorithm trait
+pub trait CoordinatorAlgorithm {
+    fn generate_sbtc_wallet_public_key(&self) -> Result<PublicKey, SBTCError>;
+    fn run_signing_round(
+        &self,
+        _tx: BitcoinTransaction,
+    ) -> Result<(Signature, SchnorrProof), SBTCError>;
+}
+
+/// Signer algorithm trait
+pub trait SignerAlgorithm: CoordinatorAlgorithm {
+    fn sign_message(&self, message: &[u8]) -> Result<Vec<u8>, SBTCError>;
+    fn verify_message(
+        &self,
+        public_key: &ecdsa::PublicKey,
+        message: &[u8],
+    ) -> Result<bool, SBTCError>;
+}
+
+/// sBTC compliant Signer
+pub struct Signer<S: SignerAlgorithm> {
+    /// Signer configuration
+    pub config: Config,
+    /// Signer private key
+    pub private_key: PrivateKey,
+    /// Network to use
+    pub network: Network,
+    /// The stacks node RPC URL
+    pub stacks_node_rpc_url: Url,
+    /// The signing algorithm
+    pub algorithm: S,
+}
+
+impl<S: SignerAlgorithm> Signer<S> {
+    /// Create a new signer
+    pub fn new(
+        config: Config,
+        private_key: PrivateKey,
+        network: Network,
+        stacks_node_rpc_url: Url,
+        algorithm: S,
```

Comment:
It might be more idiomatic to pass `S` as a generic.

```
pub fn new<S>(...) -> Self;

...

let signer: SignerAlgorithm<MegaSigner> = ...;
```

Reply:
Not sure how to do this when its already a generic stored as a field in the struct....

Reply:
Actually, ignore what I wrote for this one. When a struct has a generic type I usually expose it as a generic parameter of the function, not as an argument. If both works then it's just a matter of preference and I don't really have a strong one here.